include/ChimeraTK/ControlSystemAdapter/MappedImage.h

```diff
@@ -0,0 +1,306 @@
+  }
+
+  template<class StructHeader>
+  MappedStruct<StructHeader>::MappedStruct(
+      ChimeraTK::OneDRegisterAccessor<unsigned char>& accToData, InitData doInitData)
+  : _containerImpl(ContainerImpl::Accessor), _accToData(accToData) {
+    helpInit(doInitData == InitData::Yes);
+  }
+
+  template<class StructHeader>
+  unsigned char* MappedStruct<StructHeader>::data() {
+    switch(_containerImpl) {
+      case ContainerImpl::Accessor:
+        return _accToData.data();
+      case ContainerImpl::Vector:
+        return _vectorToData->data();
+      case ContainerImpl::CArray:
+        return _cArrToData;
+    }
+    // just to get rid of compiler warnings
+    assert(false);
+    return nullptr;
+  }
+
+  template<class StructHeader>
+  size_t MappedStruct<StructHeader>::capacity() const {
+    switch(_containerImpl) {
+      case ContainerImpl::Accessor:
+        // reason for cast: getNElements not declared const
+        return const_cast<MappedStruct*>(this)->_accToData.getNElements();
+      case ContainerImpl::Vector:
+        return _vectorToData->capacity();
+      case ContainerImpl::CArray:
+        return _cArrLenth;
+    }
+    // just to get rid of compiler warnings
+    assert(false);
+    return 0;
+  }
+
+  template<class StructHeader>
+  void MappedStruct<StructHeader>::helpInit(bool doInitData) {
+    static_assert(std::is_base_of<OpaqueStructHeader, StructHeader>::value,
+        "MappedStruct expects StructHeader to implement OpaqueStructHeader");
+    if(doInitData) {
+      if(capacity() < sizeof(StructHeader)) {
+        throw logic_error("buffer provided to MappedStruct is too small for correct initialization");
+      }
+
+      auto* p = data();
+      memset(p, 0, capacity());
+      new(p) StructHeader;
+      header()->totalLength = sizeof(StructHeader); // minimal length, could be larger
+    }
+    else {
+      if(header()->totalLength > capacity()) {
+        throw logic_error("buffer provided to MappedStruct is too small for assumed content");
+      }
+    }
+  }
+
+  /*************************** begin MappedImage implementations  ************************************************/
+
+  inline size_t MappedImage::formatsDefinition(
+      ImgFormat fmt, unsigned width, unsigned height, unsigned& channels, unsigned& bpp) {
+    switch(fmt) {
+      case ImgFormat::Unset:
+        assert(false && "ImgFormat::Unset not allowed");
+        break;
+      case ImgFormat::Gray8:
+        channels = 1;
+        bpp = 1;
+        break;
+      case ImgFormat::Gray16:
+        channels = 1;
+        bpp = 2;
+        break;
+      case ImgFormat::RGB24:
+        channels = 3;
+        bpp = 3;
+        break;
+      case ImgFormat::RGBA32:
+        channels = 4;
+        bpp = 4;
+        break;
+    }
+    return sizeof(ImgHeader) + (size_t)width * height * bpp;
+  }
+
+  inline size_t MappedImage::lengthForShape(unsigned width, unsigned height, ImgFormat fmt) {
+    unsigned channels;
+    unsigned bpp;
```

Comment:
Is this "bytes per pixel"? Is it a common abbreviation? It does not follow the coding style of "speaking variable names"

Reply:
renamed it to bytesPerPixel (since it could also be bits per pixel if you don't know)